duckduckgo/iOS
# Make SyncQueue serial and suspensible

Task/Issue URL: https://app.asana.com/0/0/1204720532613165/f

**Description**:
When app enters background, sync service will now cancel all sync operations currently in the queue and suspend
sync queue so that it doesn't run any subsequently enqueued operations. A background task is created
to ensure the app has enough time to clean up after interrupted sync.
When app enters foreground, sync service resumes sync queue and runs any previously enqueued operations,
with the caveat that those operations that have been cancelled while going to background are no-ops.



**Steps to test this PR**:
1. Enable sync (1 device is enough).
2. Make sure you have a bookmark.
3. In iOS device Developer settings, enable Network Link Conditioner and set preset to Very Bad Network
4. In the app, edit the bookmark and immediately put the app to background
5. Watch logs marked with `[DDG Sync]` and confirm that sync is cancelled and queue is suspended, and then the background task is ended when sync response is received
6. Bonus points for simulating background task timeout (100% Loss preset didn't work for me, but I have a custom preset with 14s DNS latency that worked) and verifying that the task is forcibly ended.



**Copy Testing**:


**Orientation Testing**:


**Device Testing**:


**OS Testing**:


**Theme Testing**:


---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
